rfcs/0188-batched-label-to-task-status.md

```diff
@@ -0,0 +1,83 @@
+# RFC 188 - Adding a batched "label to task status" endpoint
+* Comments: [#188](https://github.com/taskcluster/taskcluster-rfcs/pull/188)
+* Proposed by: @Alphare
+
+# Summary
+
+Add an endpoint to query the latest status for a given task label, for each
+label in a provided list.
+
+## Motivation
+
+When looking at profiles with the following command in the `mozilla-unified`
+repo:
+
+`py-spy record -F --idle --format speedscope -o output.json -- ./mach taskgraph morphed -p taskcluster/test/params/mc-onpush.yml`
+
+I saw that the majority (near 70%) of the time was spent on the optimization
+step, as the screenshot below shows:
+
+![20231211_10h54m47s_grim](https://github.com/taskcluster/taskcluster/assets/9445758/62c400cc-a125-4f08-b7dd-c8bc9a9e9a6d)
+
+Further investigating, it turns out that basically all of this time is spent
+waiting on the network doing `read()` calls. Basically, for every task, there
+is one call to `find_task_id`, then one call to `status_task`, the goal
+of which is to get the status of the latest task that corresponds to a label.
+In our case, this corresponds to 900+ API calls, incurring a massive performance
+cost to the client, and I'm guessing it's not negligible from the server's
+standpoint. (Why some of those calls take multiple seconds I have no idea,
+it might be worth investigating separately.)
+
+## Proof of concept
+
+To see if my intuition was correct, I've recorded the mapping of
+`label -> status` and replayed the command with the mapping acquired from a
+JSON file to simulate a (very) fast single API call that would give the
+right structure. The optimization step goes down to 2 seconds from 2+ minutes,
+as shown in the screenshot below:
+
+![20231211_10h55m01s_grim](https://github.com/taskcluster/taskcluster/assets/9445758/1849c8a1-fcc0-403b-acaf-ea997c875505)
+
+# Details
+
+Implementation details:
+
+- New API endpoint `/status`
+- HTTP GET:
+    - Request body consisting of a JSON object:
+        ```
+        {
+            "labels": <list of labels, not sorted, non-unique>
+        }
+        ```
+    - Response body:
+        ```
+        {
+            "mapping": {
+                <label>: <status of latest task for label>
+            },
+            "continuationToken": <continuation token>
+        }
+        ```
+
+This changes the behavior of the program slightly since the task statuses
+are queried all at once and not during the optimization process. This is not
+a problem since the same "race condition" of observing a new status for
+any given task exists in the current code, just much slower. I would argue
+that making this a single step also makes debugging easier.
+
+There are no compatibility or security concerns.
+
+I'm a little blurry on how the database is currently laid out, but I've had
```

Comment:
So I can give a couple of insights. 
At the moment this information is stored in a `tasks` table which has about 28.900.000 records at the time of writing. (160GB data, 9GB indexes).

`"label"` is not part of the "core" (not sure if this is a correct wording :) schema, and is stored in the `task.tags` column as JSONB dictionary. To query it in a way that wouldn't including scanning all 29kk records on disk would probably involve indexes.

Scanning all records takes about 4 minutes at the moment, so we cannot do that. 

The only two indexes that exist on that table is one for `task_id` and other for `task_group_id`

My concern at this point is that `task.tags` is a  "free-form" dictionary provided to store extra information, and having dedicated endpoint just for one of the possible values there might not be the best way of doing this. 

I'm wondering if we should use namespaces/indexes to query for the latest task with that label, but the problem there could be that task is only added to namespace once it has finished.

Reply:
I'll resolve this thread, it seems like what actually is needed is something else.
taskgraph is querying some indexes and for those indexes fetches task statuses